doc/source/how-to-use-driver-api.rst

```diff
@@ -0,0 +1,254 @@
+Use Driver API
+==============
+
+This guide explains how you can use Flower framework's |driverapi_link|_ API to customize the :code:`ServerApp` for *any* server workloads in your federated learning system. The Driver APIs are a specific set of server-side APIs that provide a robust way to communicate with client nodes in a Flower federated learning system. They include methods to:
+
+* `Select nodes`_
+* `Push messages to nodes`_
+* `Pull messages from nodes`_
+
+..
+    Generate link text as literal.
+
+.. |driverapi_link| replace:: ``Driver()``
+.. |recordset_link| replace:: ``RecordSet()``
+.. |flower_serverapp_link| replace:: ``flower-server-app``
+.. _driverapi_link: ref-api/flwr.server.Driver.html
+.. _recordset_link: ref-api/flwr.common.RecordSet.html
+.. _flower_serverapp_link: ref-api-cli.html#flower-server-app
+
+Setup
+-----
+
+Before diving into the :code:`Driver` API, we need to setup your :code:`server.py` module. In a file :code:`server.py`, import :code:`flwr` and create a :code:`ServerApp`:
+
+.. code-block:: python
+
+    import flwr
+    from flwr.server import ServerApp, Driver
+    from flwr.common import Context, log
+
+    # Run via `flower-server-app server:app`
+    app = ServerApp()
+
+Next, we initialize the :code:`main()` function using :code:`@app.main()`. Within :code:`main()`, we define the workload that we want. For example, in federated learning workloads, we define a number of server rounds (:code:`server_round`) and loop over them throughout training:
+
+.. code-block:: python
+
+    # File: server.py
+
+    @app.main()
+    def main(driver: Driver, context: Context):
+        num_rounds = 3
+    
+        for server_round in range(num_rounds):
+            print(f"Commencing server round {server_round + 1}")
+            
+            # Do something using Driver APIs
+
+Now that we have completed the setup, let’s walk through how we can use the Driver APIs for a federated learning example.
+
+
+Select nodes
+------------
+
+The first stage of a federated learning workflow is to select client nodes. For simplicity, we will refer to client nodes as nodes throughout this guide. 
+
+To select nodes in each server round, we run :code:`Driver.get_node_ids()` to retrieve all connected node IDs. Occasionally, the Driver API may not immediately return enough node IDs (e.g. due to real-world network issues), so we loop and wait until the minimum number of nodes are available. If we have a large pool of nodes which we want to sample from - as is commonly the case for federated learning systems - we can randomly sample node IDs using, for instance, :code:`random.sample(all_node_ids, num_client_nodes_per_round)`. One example is shown below:
```

Comment:
"real-world network issues" might be a bit vague for some readers without background about FL. A couple of examples that come to mind that we can mention: some nodes went offline, a node is taking longer than expected to startup.